### PR TITLE
[Impeller] Curve components in stroke path use start directions as their initial offsets

### PR DESCRIPTION
**Pull request by chunhtai**
The golden changes should test the change 

fixes https://github.com/flutter/flutter/issues/135225

This is what it should look like
image

but because the first point offset uses the direction to the second point, the line draws like this
image

so it miss the yellow part( the reason the missing pixel in the original issue) and overdraw the purple part
image

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by chunhtai** [comment]
hmm looks like I broke some golden again. will take a look

File: impeller/aiks/aiks_unittests.cc [hunk 1]
```diff
@@ -246,6 +246,16 @@ TEST_P(AiksTest, CanRenderCurvedStrokes) {
   ASSERT_TRUE(OpenPlaygroundHere(canvas.EndRecordingAsPicture()));
 }
 
+TEST_P(AiksTest, CanRenderThickCurvedStrokes) {
```
Comment (chunhtai) [inline impeller/aiks/aiks_unittests.cc:249]: May I have some advice? Not entirely sure why this produce clipped golden
image
locally it runs fine

File: impeller/entity/geometry/stroke_path_geometry.cc [hunk 2]
```diff
@@ -242,14 +242,99 @@ StrokePathGeometry::CreateSolidStrokeVertices(
...
+                         contour_end_point_i, contour);
+          if (!is_last_component && is_end_of_component) {
+            // Generate join from the current line to the next line.
+            join_proc(vtx_builder, polyline.points[point_i + 1],
+                      previous_offset, offset, scaled_miter_limit, scale);
+          }
+        }
+      };
+
+  auto add_vertices_for_curve_compoent =
```
Comment (chunhtai) [inline impeller/entity/geometry/stroke_path_geometry.cc:301]: There are still one more issue with curve component. the overdraw from the few start points will draw outside of the start point. I would like to address this in a later pr.
image

The start of the curve is not exactly horizontal due to this issue.

**Review by chunhtai — COMMENTED** [review]
The first commit is the straight reland of the reverted pr. The new change is in the second pr.

**Review by bdero — APPROVED** [review]
LGTM!

Apologies for the delay, this one fell off my radar during the summit.

**Comment by bdero** [comment]
Huh, I just pulled down [caed4a6](https://github.com/flutter/engine/pull/46203/commits/caed4a6f9d37231c3fd1f846aa395df99ff27447) and I can't seem to reproduce the failed golden behavior for impeller_Play_AiksTest_CanRenderCurvedStrokes_Metal. Perhaps there's a golden tracking issue.

image

**Pull request merged**